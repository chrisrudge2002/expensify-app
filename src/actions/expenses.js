import uuid from 'uuid';
import { expenseActionTypes } from './types';

export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => {
    return {
        type: expenseActionTypes.ADD_EXPENSE,
        expense: {
            id: uuid(),
            description,
            note,
            amount,
            createdAt
        }
    };
};

export const editExpense = (id, updates = {}) => {
    return {
        type: expenseActionTypes.EDIT_EXPENSE,
        id,
        updates
    };
};

export const removeExpense = ({ id } = {}) => {
    return {
        type: expenseActionTypes.REMOVE_EXPENSE,
        id
    };
};