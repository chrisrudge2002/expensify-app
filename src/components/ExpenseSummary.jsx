import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeraljs';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expenseCount, totalAmount }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedTotal = numeral(totalAmount / 100).format('$0,0.00') 
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedTotal}</h1>
        </div>
    );
};


const mapStateToProps = (state) => {
    return {
        expenseCount: selectExpenses(state.expenses, state.filters).length,
        totalAmount: selectExpensesTotal(selectExpenses(state.expenses, state.filters))
    }
}

export default connect(mapStateToProps)(ExpenseSummary);