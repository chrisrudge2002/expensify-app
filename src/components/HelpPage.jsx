import React from 'react';

export default () => (
    <div>HelpPage</div>
)
