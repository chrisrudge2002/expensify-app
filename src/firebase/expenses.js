import  database from './firebase';

export const addExpenseToDatabase = async (expense) => {
    const result = await database.ref('expenses').push(expense);
    return {
        ...expense,
        id: result.key
    };    
};

export const editExpenseInDatabase = async (id, updates) => {
    await database.ref(`expenses/${id}`).update({...updates});    
};

export const removeExpenseFromDatabase = async (id) => {
    await database.ref(`expenses/${id}`).remove();
}

export const getExpensesFromDatabase = async () => {
    const snapshot = await database.ref('expenses').once('value');
    const expenses = [];
    snapshot.forEach((item) => {
        expenses.push({
            ...item.val(),
            id: item.key
        });
    });
    return expenses;
};