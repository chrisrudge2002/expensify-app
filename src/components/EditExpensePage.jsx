import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import * as expenseActions from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                    <button 
                        className="button button--secondary" 
                        onClick={this.onRemove}>
                            Remove Expense
                    </button>
                </div>
            </div>
        );
    };
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(e => e.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(expenseActions.startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(expenseActions.startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);