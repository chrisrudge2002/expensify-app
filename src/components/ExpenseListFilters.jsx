import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import * as filterActions from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onCalendarFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            this.props.sortByAmount();
        }
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input"
                            type="text" 
                            value={this.props.filters.text} 
                            onChange={this.onTextChange} placeholder="Search expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy} 
                            onChange={this.onSortChange}>
                                <option value="date">Date</option>
                                <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                        startDate={this.props.filters.startDate}
                        startDateId="startDate"
                        endDate={this.props.filters.endDate}
                        endDateId="endDate"
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                        displayFormat="DD MMM YYYY"
                        showClearDates={true}
                    />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setEndDate: (endDate) => dispatch(filterActions.setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(filterActions.setStartDate(startDate)),
    setTextFilter: (text) => dispatch(filterActions.setTextFilter(text)),
    sortByAmount: () => dispatch(filterActions.sortByAmount()),
    sortByDate: () => dispatch(filterActions.sortByDate()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);