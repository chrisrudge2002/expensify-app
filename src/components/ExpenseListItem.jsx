import React from 'react';
import { Link } from 'react-router-dom';
import * as expenseActions from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{amount} - {createdAt}</p>
    </div>
);

export default ExpenseListItem;