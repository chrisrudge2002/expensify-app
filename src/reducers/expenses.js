import { expenseActionTypes } from '../actions/types';

const defaultState = [];

export default (state = defaultState, action) => {
    switch (action.type) {
        case expenseActionTypes.ADD_EXPENSE: 
            return [
                ...state,
                action.expense
            ];
        case expenseActionTypes.EDIT_EXPENSE: 
            return state.map(e => {
                if (e.id === action.id) {
                    return {
                        ...e,
                        ...action.updates
                    };
                }
                return e;
            });
        case expenseActionTypes.REMOVE_EXPENSE: 
            return state.filter(e => e.id !== action.id);
        default: 
            return state;
    }
};

