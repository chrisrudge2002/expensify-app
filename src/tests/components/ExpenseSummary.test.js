import React from 'react';
import { shallow }  from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render ExpenseSummary correctly with no expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={0} totalAmount={0} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly with one expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} totalAmount={4500} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={2} totalAmount={1234500} />);
    expect(wrapper).toMatchSnapshot();
});