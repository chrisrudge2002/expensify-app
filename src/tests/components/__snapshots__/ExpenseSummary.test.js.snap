// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    $12,345.00
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly with no expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $45.00
  </h1>
</div>
`;
